### DISCUSSION TITLE
Weird behaviour of agents with constant target

### DISCUSSION THREAD

**Discussion by FrameMuse** [isaac-mason/recast-navigation-js#395] [General]
I'm having an issue whenever I unfocus the tab where my game is running, units spreading across the map randomly. But when I focus the tab again they start to get to their position again.

I have it very exaggerated, but [the demo](https://recast-navigation-js.isaacmason.com/?path=/story/crowd-crowd-with-multiple-agents--crowd-with-multiple-agents) also has the problem.

Just after focusing
![image](https://github.com/isaac-mason/recast-navigation-js/assets/53980482/49863f4f-2fbc-43cd-a5db-f34605e8dcf0)

After some time
![image](https://github.com/isaac-mason/recast-navigation-js/assets/53980482/35ffa18c-c8c1-4700-845c-3370f9f1cfa8)

**Comment by isaac-mason** [comment]
Hi @FrameMuse, I believe this issue is caused by calling crowd.update with a large delta time.

This should be addressed (in the examples too!) by either, clamping the delta time to some value, e.g. Math.min(delta, 0.1), or by pausing calling update based on page visibility.

**Reply by FrameMuse** [reply]
Yes, you are right with the problem. When I put a constraint `Math.min(delta, 0.1)`, it's much better, but now when I focus the page, the agents "shake" and "jump" a little bit, which doesn't really look good. Imagine a group of heroes **tremble** when your game is lagging a little bit.

So probably the best solution _right now_ is to implement game pausing whenever user is out of **canvas**, so it doesn't calculate anything and let user continue when they are there.

But I would love to have a "fix" for this since I know many games don't have that problem or maybe someone else could come with a solution for current API - it would be great.

---

[Page visibility API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) is not perfect as well, sometimes it doesn't really triggers page hiding even though it is.

**Reply by isaac-mason** [reply]
Hey @FrameMuse, some combination of clamping the delta time and the page visibility API sounds like a good idea.

Calling crowdAgent.resetMoveTarget when the agent is sufficiently close to the target may help with this situation too.

You could also try doing fixed time stepping with interpolation. This would mean in the worst case where the visibility API doesn't work correctly multiple stable fixed sized steps would be performed.

See the "Fixed time stepping with interpolation" section in the readme 🙂

**Reply by GGAlanSmithee** [reply]
I whish Chromium would allow you to run RAF in a non-focused tab. I understand the reason to not, but they could allow it behind user consent (like audio, etc). In my experience, it is _very_ hard to get this perfect without it, because there will always be some form of discrepancy when re-focusing.

In theory, you could override / create your own RAF which is paused/unpaused when unfocusing/focusing, but as you point out yourself, it's not 100% reliable, and practically impossible to monkey patch third party packages, like for exampl r3f's `useFrame`, etc.

Electron does allow for this with `backgroundThrottling: false`, but this is unfortunately not possible in a pure web context.

**Reply by FrameMuse** [reply]
@isaac-mason Yes, I investigated this approach just before you mentioned it. It seems it makes it MUCH reliable.

Thanks @GGAlanSmithee for inspiration, now I'd like to experiment with background-sync, Service Workers and Web Workers to see if I can run the game there smoothly there. Or at least dedicate some logic of synchronization there (not sure though, that's just an idea).

I will share everything when I finish with all my investigations.

**Reply by GGAlanSmithee** [reply]
Nice @FrameMuse!

Just to add to the discussion, I have done some experiments with using Web Workers in the context of three.js and recast-navigation, but often the overhead of passing buffers and other information back and forth negates most of the potential wins. However, I didn't really do a deep enough dive for anything conclusive, so would be nice to see what you come up with!

**Reply by GGAlanSmithee** [reply]
I recently had a very positive experience moving some navigation mesh creating code over to a worker (copied from one of the examples), as I explained here: https://github.com/isaac-mason/recast-navigation-js/discussions/405
